gnovm/pkg/gnolang/preprocess.go

```diff
@@ -882,17 +882,22 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 					// Left not const, Right not const ------------------
 					if n.Op == EQL || n.Op == NEQ {
 						// If == or !=, no conversions.
-					} else if lnt, ok := lt.(*NativeType); ok {
+					} else if lnt, ok := lt.(*NativeType); ok && isNative(rt) {
 						if debug {
 							if !isShift {
 								assertSameTypes(lt, rt)
 							}
 						}
-						// If left and right are native type,
+						// If left and right are native type, and same type
 						// convert left and right to gno, then
 						// convert result back to native.
 						//
 						// get concrete native base type.
+						if lt.TypeID() != rt.TypeID() {
+							panic(fmt.Sprintf(
+								"incompatible types in binary expression: %v %v %v",
+								n.Left, n.Op, n.Right))
```

Comment:
maybe print the type info is more straightforward, what do you think?

Reply:
And, can you add a file test to better display this error message?

Reply:
Yes right, I will do it

Reply:
Added one more test here: https://github.com/gnolang/gno/pull/2016/commits/e6e8d333c593deb210f812f4f5770a0fd65b7e92

> maybe print the type info is more straightforward, what do you think?

What do you mean here ?

Reply:
```go
panic(fmt.Sprintf("incompatible types in binary expression: %v %v %v",lt.TypdID(), n.Op, rt.TypeID()))
															
```
what do you think?

Reply:
I don't have an opinion on this. I don't know why we did it before (but the typeID printout is clearer).

Done here: https://github.com/gnolang/gno/pull/2016/commits/959b43e5bc7ff886586e499262b76644a8e0b1df